### PR TITLE
fix missing "back" links in coach 

### LINKED ISSUE #10673: Missing 'All classes' link on device with multiple facilities
## Observed behavior
The 'All classes' link is missing on a device with multiple facilities so once I have selected one of the available classes I have to navigate to a different plugin in order to go back to the Classes page.

## Expected behavior
The  'All classes' link should be displayed in that case.

general expectations for `Coach` behavior & UI:

- if **user 1** is a multi-facility admin of Facility A & Facility B, and Facility A has only one class, `Class home` goes directly to that class's page and when there, the back link is "← Change learning facility"
- when **user 1** selects Facility B, which has multiple classes, they are redirected to the `All Classes` page which has a back link to "← Change learning facility".
    - selecting Class B1 will take them to its specific class page, which contains a back link to "← All classes"
- if **user 2** has access to only Facility A, which has one class (or user has other limitations only allowing access to one class), `Class home` goes directly to that class's page and there is no back link displayed.
- if **user 3** has access to only Facility B, which has multiple classes, they are redirected to the `All Classes` page and there is no back link displayed.
   - selecting Class B1 will take you to its specific class page with a back link to "← All classes"

## Steps to reproduce the issue
1. Install the the build asset from [this PR](https://github.com/learningequality/kolibri/pull/10550)
2. Import a facility with multiple classes
3. Navigate to Coach > Class home and select a class
4. Attempt to go back to the Classes page

Note: observed in edge case when multi-facility user's main facility has 1 class and imported facility has multiple, detailed in comments

## Video

https://github.com/learningequality/kolibri/assets/79847249/a1d02b87-1739-4c01-83a2-a831a1341164

## Usage Details
Ubuntu 22  - Chrome, Firefox

### PR DESCRIPTION
**Pull request by thanksameeelian**
## Summary
as reported in #10673, the back link "← All classes" was not displaying as expected when a multi-facility admin was exploring classes within an imported facility **_if_** their default facility had only 1 class. 

though this was an edge case, it uncovered a pervasive issue with the `setClassList` action, which is used throughout the `Coach` app to determine the list of classes to display & decide which back links to render in which cases. when `setClassList` was called without a `facilityId` argument, it used the value of `currentFacilityId` to inform its behavior, and that value always corresponds to a multi-facility user's default facility. though it only became evident in these specific conditions, we were in fact generally assessing the wrong facility much of the time we invoked `setClassList` in an imported facility.

this PR introduces changes to `setClassList` and how it is used - now, it requires a `facilityId` argument and throws an error if one is not provided. in each place the action is called, it is now supplied with a `facilityId` that has been accessed through `store.state.classSummary`, which is also now guaranteed to be set wherever `setClassList` is invoked.

## References
fixes #10673

## Reviewer guidance
**Note:** this bug appeared only in the special edge case that the user was a multi-facility admin, with 1 class in their default facility & multiple classes in an imported facility. please replicate that scenario to test these changes, but also keep in mind the full list of expectations outlined in #10673 & investigate as much as is reasonable:
- if user 1 is a multi-facility admin of Facility A & Facility B, and Facility A has only one class:
    - selecting Facility A will lead the user directly to the sole class's `Class home` page. when there, the back link is "← Change learning facility"
    - if user navigates to `Plan` or `Report` subtopics, the back link "← Class home" will be displayed [_NEW & open to debate - added for ease of navigation back to `All facilities` page in this specific multi-facility, single-class scenario_]
- when user 1 selects Facility B, which has multiple classes:
    - they are redirected to the `All Classes` page which has a back link to "← Change learning facility"
    - selecting a class from `All Classes` will take the user that class's specific `Class home` page, which contains a back link to "← All classes"
    - once within a specific class, if user navigates to `Plan` or `Report` subtopics, the back link "← Class home" will be displayed
- if user 2 has access to only Facility A, which has one class (or user has other limitations only allowing access to one class):
    - selecting `Class home` within `Coach` in the top- or sidenav leads directly to that class's page
    - there are no back links displayed
- if user 3 has access to only Facility B, which has multiple classes:
    - selecting `Class home` within `Coach` in the top- or sidenav redirects them to the `All Classes` page and there is no back link displayed.
    - selecting a class will take user to its specific class page with a back link to "← All classes"
    - once within a specific class, if user navigates to `Plan` or `Report` subtopics, the back link "← Class home" will be displayed

changes introduced in this PR impacted the code wherever `setClassList` was invoked, so while exploring the scenarios outlined above, it would also be helpful to visit the two additional places where this occurs, which are encountered when digging into the details of a specific class:
- `LessonSummaryPage`, found at `/:classId/plan/lessons/:lessonId` for a specific lesson
- `CoachExamsPage`, found at `/:classId/plan/quizzes`

----

## Testing checklist

- [X] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [X] PR has the correct target branch and milestone
- [X] PR has 'needs review' or 'work-in-progress' label
- [X] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/plugins/coach/assets/src/modules/pluginModule.js [hunk 1]
```diff
@@ -60,13 +60,17 @@ export default {
...
+      if (facilityId) {
+        return ClassroomResource.fetchCollection({
+          getParams: { parent: facilityId, role: 'coach' },
         })
-        .catch(error => store.dispatch('handleApiError', error));
+          .then(classrooms => {
+            store.commit('SET_CLASS_LIST', classrooms);
+          })
+          .catch(error => store.dispatch('handleApiError', error));
+      } else {
```
Comment (rtibbles) [inline kolibri/plugins/coach/assets/src/modules/pluginModule.js]: As you are returning in the above `if` statement, the `else` here is not needed!

Alternatively, you could do an 'error early' strategy here and do a `!facilityId` check throw the error, and then just carry on with the unindented function.

**Review by rtibbles — COMMENTED** [review]
All code changes make sense to me! One comment, but definitely not a blocker.

File: kolibri/plugins/coach/assets/src/modules/pluginModule.js (hunk 1 above)
Comment (thanksameeelian) [reply kolibri/plugins/coach/assets/src/modules/pluginModule.js]: updated to match your second suggestion!

**Comment by pcenov** [comment]
Hi @thanksameeelian I've started testing this today and so far it's looking great when there are multiple facilities. However if there aren't any imported facilities one is not able to access the Coach plugin - getting the following error in the console "Error: Missing required 'facilityId' argument"

**Comment by rtibbles** [comment]
Seems like exactly the sort of error we wanted to throw! Seems like maybe there's a case where the facilityId is being passed to the action, but it is null or undefined?

**Comment by pcenov** [comment]
Hi @thanksameeelian the only other issue which I was able to spot is a brief display of the 'Classes' page prior to navigating to the 'Class home' page. This is best observed with the browser's throttling option set to 'Slow 3G':

https://github.com/learningequality/kolibri/assets/79847249/ad0cd2ed-51c2-4264-93c5-b0ba04ba6786

Let me know when you've fixed these 2 issues so I can fully regression test everything.

**Comment by rtibbles** [comment]
@pcenov is the issue you are seeing this one? [#10714](https://github.com/learningequality/kolibri/issues/10714)

**Comment by thanksameeelian** [comment]
> Hi @thanksameeelian the only other issue which I was able to spot is a brief display of the 'Classes' page prior to navigating to the 'Class home' page. This is best observed with the browser's throttling option set to 'Slow 3G':
> 
>  2023-06-06_12-06-18.mp4 
> Let me know when you've fixed these 2 issues so I can fully regression test everything.

yes @pcenov, as @rtibbles mentioned, [this is a bug that i reported 3 weeks ago that is unrelated to my changes](https://github.com/learningequality/kolibri/issues/10714) which i also detailed [in a comment on jacob's coach loading states PR](https://github.com/learningequality/kolibri/pull/10455#discussion_r1205881575), and it will be getting addressed separately. please let me know if there's anything related to this PR that still needs fixing up though! thank you for being so thorough!

**Comment by pcenov** [comment]
Yeah, @thanksameeelian... I thought I've seen that somewhere but having being pressed by time didn't get to check for existing issues so just mentioned it here. Otherwise in case you have missed my first comment - that now it's not possible to access the Coach plugin if you have only one Facility - is the main and only issue that needs to be fixed so that we can finalize this task.

**Comment by thanksameeelian** [comment]
ah, thank you so much @pcenov! i must admit that i actually (thought i had) caught up on these comments through reading email notifications in my inbox 🫣, and missed a little bit of the back and forth about that initial error and somehow thought it had been determined to be a non-issue. i'm sorry to make you repeat yourself - i'll get working on fixing this up 🙂

(also i believe @nucleogenesis made a small change or two while he was tinkering around in coach for another purpose, so once i also resolve the merge conflict here, i think the weird issue of that brief display _may_ be somewhat mitigated, but definitely still extant)

**Comment by thanksameeelian** [comment]
@pcenov [the bug you mention here](https://github.com/learningequality/kolibri/pull/10768#issuecomment-1577001589) should now be addressed & working smoothly!

**Comment by pcenov** [comment]
Hi @thanksameeelian I confirm - this issue is fixed now and everything else is working as specified.

**Review by radinamatic — APPROVED** [review]
Manual QA passes, good to go! 💯 👏🏽 :shipit:

**Review by rtibbles — APPROVED** [review]
